### PR TITLE
Humans and horses

### PR DESCRIPTION
**Pull request by msluszniak**

### CODE REVIEW

File: notebooks/humans_or_horses.livemd [hunk 1]
```diff
@@ -0,0 +1,161 @@
+# Humans or horses
+
+## General overview
+
+In this notebook, we want to predict wheater an image present a human or a horse. To do this efficiently, we will build a convolutional neural net and compare the learning process with and without gradient centralization.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:5]:
```suggestion
In this notebook, we want to predict whether an image presents a human or a horse. To do this efficiently, we will build a convolutional neural net and compare the learning process with and without gradient centralization.
```

File: notebooks/humans_or_horses.livemd [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+## Imports
+
+```elixir
+Mix.install([
+  {:stb_image, "~> 0.1.0"},
+  {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:12]:
```suggestion
  {:axon, "~> 0.1.0"},
```

File: notebooks/humans_or_horses.livemd [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+EXLA.set_as_nx_default(
+  [:tpu, :cuda, :rocm, :host],
+  run_options: [keep_on_device: true]
+)
+```
+
+## Data processing
+
+Firstly, we need to preprocess the data to our CNN. At the beginning of the process, chunk images into batches. We use the ```parse_png``` function to load images and label them accurately. In the end, we "augment" the input, which means that we normalize data and flip the images along one of the axes. The last procedure helps NN to make predictions regardless of the orientation of the image.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:25]:
```suggestion
First, we need to preprocess the data for our CNN. At the beginning of the process, we chunk images into batches. We use the ```parse_png``` function to load images and label them accurately. Finally, we "augment" the input, which means that we normalize data and flip the images along one of the axes. The last procedure helps neural network to make predictions regardless of the orientation of the image.
```

File: notebooks/humans_or_horses.livemd [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+<!-- livebook:{"break_markdown":true} -->
+
+| ![](https://miroslawmamczur.pl/wp-content/uploads/2021/03/06.gif)         |
+| :-----------------------------------------------------------------------: |
+| Figure 1: An example of convolutional layer works for *kernel_size*={3,3} |
+
+<!-- livebook:{"break_markdown":true} -->
+
+```Axon.max_pool``` is a mathematical operation which gives maximum value from a subtensor of a size equal to *kernel_size*.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:85]:
```suggestion
```Axon.max_pool``` is a mathematical operation which gives the maximum value from a subtensor of a size equal to *kernel_size*.
```

File: notebooks/humans_or_horses.livemd [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    |> Nx.from_binary({:u, 8})
+    |> Nx.reshape(shape)
+
+  {tensor, class}
+end
+```
+
+## Building Model
+
+The next step is creating a model architecture. In the notebook, we choose classic Convolutional Neural Net architecture. There are a few Axon functions that I will try to explain briefly.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:71]:
```suggestion
The next step is creating a model architecture. In the notebook, we choose classic Convolutional Neural Net architecture. Let's dive in to the core components of a CNN.
```

File: notebooks/humans_or_horses.livemd [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+<!-- livebook:{"break_markdown":true} -->
+
+| ![](https://production-media.paperswithcode.com/methods/MaxpoolSample2.png) |
+| :-------------------------------------------------------------------------: |
+| Figure 2: Max pooling operation for *kernel_size*={2,2}                     |
+
+<!-- livebook:{"break_markdown":true} -->
+
+Axon.dropout and Axon.spatial_dropout creates dropout layers which prevent NN from overfitting. Standard dropout drops a given rate of randomly chosen neurons during the training process. On the other hand, spatial dropout gets rid of whole feature maps. The graphical difference between dropout and spatial dropout is presented in a picture below.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:95]:
```suggestion
Axon.dropout and Axon.spatial_dropout create dropout layers which prevent the neural network from overfitting. Standard dropout drops a given rate of randomly chosen neurons during the training process. On the other hand, spatial dropout gets rid of whole feature maps. The graphical difference between dropout and spatial dropout is presented in a picture below.
```

File: notebooks/humans_or_horses.livemd [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  model
+  |> Axon.Loop.trainer(:binary_cross_entropy, optimizer, log: 1)
+  |> Axon.Loop.metric(:accuracy)
+  |> Axon.Loop.run(data, %{}, epochs: epochs, iterations: 100)
+end
+```
+
+## Model evaluation
+
+Now we can build and train our model. We start with passing input shape and indices to data transposition. Next, we set the optimizer to adam with a learning rate of 1.0e-4. In the next step, we define an optimizer but with gradient centralization. It is technic with a similar purpose to batch normalization. For each loss gradient, we subtract a mean value to have a gradient with mean equal to zero. This process prevents us from exploding gradients.
```
Comment (seanmor5) [inline notebooks/humans_or_horses.livemd:145]:
```suggestion
Now we can build and train our model. We start with passing input shape and indices to data transposition. Next, we set the optimizer to adam with a learning rate of 1.0e-4. In the next step, we define an optimizer but with gradient centralization. It is a technique with a similar purpose to batch normalization. For each loss gradient, we subtract a mean value to have a gradient with mean equal to zero. This process prevents us from exploding gradients.
```

**Review by seanmor5 — APPROVED** [review]
Awesome! I just made a few minor edit suggestions, once they are addressed it's good to merge :)
